xrootd/encoder/encoder.go

```diff
@@ -0,0 +1,178 @@
+func Marshal(x interface{}) ([]byte, error) {
+	v := reflect.ValueOf(x)
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+
+	if v.Kind() != reflect.Struct {
+		return nil, errors.Errorf("Cannot marshal %s, struct is expected", v.Kind())
+	}
+
+	dataSize, err := calculateSizeForMarshaling(v)
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, dataSize)
+	pos := 0
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
+		case reflect.Uint8:
+			fieldSize = 1
+			data[pos] = uint8(field.Uint())
+		case reflect.Uint16:
+			fieldSize = 2
+			binary.BigEndian.PutUint16(data[pos:pos+fieldSize], uint16(field.Uint()))
+		case reflect.Int32:
+			fieldSize = 4
+			binary.BigEndian.PutUint32(data[pos:pos+fieldSize], uint32(field.Int()))
+		case reflect.Int64:
+			fieldSize = 8
+			binary.BigEndian.PutUint64(data[pos:pos+fieldSize], uint64(field.Int()))
+		case reflect.Slice:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal slice of %s, only slices of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
+		case reflect.Array:
+			elementKind := field.Type().Elem().Kind()
+			if elementKind != reflect.Uint8 {
+				return nil, errors.Errorf("Cannot marshal array of %s, only arrays of uint8 are supported", elementKind)
+			}
+
+			fieldSize = field.Len()
+			reflect.Copy(reflect.ValueOf(data[pos:pos+fieldSize]), field)
+
+		default:
+			return nil, errors.Errorf("Cannot marshal kind %s", field.Kind())
+		}
+		pos += fieldSize
+	}
+
+	return data, nil
+}
+
+// Unmarshal decodes data from byte slice following the XRootd protocol specification.
+// Fields are decoded in the same order as they are specified in the struct definition.
+// Each field is decoded from network byte order (BigEndian) without any alignment or padding.
+// Slices and arrays of uint8 are binary copied without further decoding.
+// Since the length of the slice is unknown, all bytes to the end of data is copied to it.
+// Supported types are: uint8, uint16, int32, int64, slices and arrays of uint8.
+func Unmarshal(data []byte, x interface{}) (err error) {
+	v := reflect.ValueOf(x)
+
+	if v.Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+
+	if v.Kind() != reflect.Struct {
+		return errors.Errorf("Cannot unmarshal %s, struct is expected", v.Kind())
+	}
+
+	pos := 0
+
+	for i := 0; i < v.NumField() && err == nil; i++ {
+		field := v.Field(i)
+		fieldSize := 0
+		switch field.Kind() {
+		case reflect.Uint8:
+			fieldSize = 1
+			field.SetUint(uint64(data[pos]))
+		case reflect.Uint16:
+			fieldSize = 2
+			var value = binary.BigEndian.Uint16(data[pos : pos+fieldSize])
+			field.SetUint(uint64(value))
+		case reflect.Int32:
+			fieldSize = 4
+			var value = int32(binary.BigEndian.Uint32(data[pos : pos+fieldSize]))
```

Comment:
why do you add this `uint32 -> int32` conversion when you convert `value` to an `int64` just the line after?
is it really needed?